### ISSUE TITLE
Disabled API key => new authentication via OAuth

### ISSUE THREAD

**Issue by hracik** [sverraest/revolut-php#9]
Revolut disabled API keys on 3 February 2020. Will there be an update for this package to authenticate via OAuth?

**Comment by vdbelt** [comment]
+1 for OAuth/JWT helpers (can help with PR if needed, @sverraest).

Currently you need to generate the token yourself (libraries as `firebase/php-jwt` can help with that).
The access token you get back from the `/api/1.0/auth/token` endpoint can then be used as a regular token when instantiating a new `RevolutPHP\Client`.


https://revolutdev.github.io/business-api/#revolut-api-authentication

**Comment by tbclla** [comment]
I had the same issue and have extended RevolutPHP in my project to handle the new JWT authentication.
I will leave my (minor) changes and additions here for anyone else who needs a solution before this package is updated.

In order to use it, you will need to get an access token (previously ApiKey) and use this to new up your RevolutPHP\Client:

```php
$accessToken = AccessToken::request($clientId, $privateKey, $issuer, $authToken);

$revolut = new RevolutPHP\Client($accessToken);
```

To make this work, you would need to make two small changes to your RevolutPHP\Client, and add a couple of new classes.

### Minor Changes to RevolutPHP/Client
Update the constructor to accept an instance of RevolutPHP\Auth\AccessToken, instead of the previous 'apiKey' string.
The 'mode' property can be removed, as the baseUrl will now depend on the environment of the access token.

```php
/**
* @var \RevolutPHP\Auth\AccessToken
*/
private $accessToken;

/**
* Client constructor.
* @param \RevolutPHP\Auth\AccessToken $accessToken
* @param array $clientOptions
*/
public function __construct(AccessToken $accessToken, array $clientOptions = [])
{
	$this->accessToken = $accessToken;
	$this->baseUrl = $accessToken->production ? self::REVOLUT_PRODUCTION_ENDPOINT : self::REVOLUT_SANDBOX_ENDPOINT;
	$this->clientOptions = $clientOptions;

	$this->initiateHttpClient();

	$this->accounts = new Accounts($this);
	$this->counterparties = new Counterparties($this);
	$this->payments = new Payments($this);
	$this->transfers = new Transfers($this);
	$this->transactions = new Transactions($this);
	$this->rates = new Rates($this);
	$this->exchanges = new Exchanges($this);
	$this->webhooks = new Webhooks($this);
}
```
then update the initiateHttpClient function to use the AccessToken's 'token' property.
```php
/**
* Initiates the HttpClient with required headers
*/
private function initiateHttpClient()
{
	$options = [
		'headers' => [
			'Authorization' => 'Bearer ' . $this->accessToken->token,
		]
	];

	$this->httpClient = new GuzzleClient(array_replace_recursive($this->clientOptions, $options));
}
```
### Additions
Create a new RevolutPHP\Auth namespace to house all the new auth related classes.

Revolut lets you retrieve an access token in two ways.
Either by submitting an authorization code or a refresh token (see documentation).

Create an abstract class that both the authorization code and the refresh token will extend:

```php
namespace RevolutPHP\Auth;

abstract class AuthToken
{
	public function name()
	{
		return $this->name;
	}

	public function grantType()
	{
		return $this->grantType;
	}
}
```
Create a class for the authorization code
```php
namespace RevolutPHP\Auth;

class AuthorizationToken extends AuthToken
{
	public $grantType = 'authorization_code';

	public $name = 'code';

	public $token;
	
	public function __construct(string $token)
	{
		$this->token = $token;
	}
}
```
Create a clas for the refresh token
```php
namespace RevolutPHP\Auth;

class RefreshToken extends AuthToken
{
	public $grantType = 'refresh_token';

	public $name = 'refresh_token';

	public $token;
	
	public function __construct(string $token)
	{
		$this->token = $token;
	}
}
```



### AccessToken
A simple class to represent an access token, containing the actual token string, the expiration date, the environment for the token(true = production, false = sandbox) and optionally a refresh token.
Includes a static function to request a new access token via a AccessTokenRequest (next section).

```php
namespace RevolutPHP\Auth;

class AccessToken
{
	/**
	 * The token string
	 * 
	 * @param string $token
	 */
	public $token;

	/**
	 * The expiration date as timestamp
	 * 
	 * @param int  $expiration
	 */
	public $expiration;

	/**
	 * Whether or not the token is for the prodcution environment
	 * 
	 * @param string  $production
	 */
	public $production;

	/**
	 * The refresh token
	 * A refresh token is only provided when exchaning an authorization code for an access token,
	 * not when exchaning a refresh token 
	 * 
	 * @param string|null $refreshToken
	 */
	public $refreshToken;

	/**
	 * @param string $token	
	 * @param int $expiration
	 * @param bool $production
	 * @param string|null $refreshToken
	 */
	public function __construct(string $token, int $expiration, bool $production, string $refreshToken = null)
	{
		$this->token = $token;

		$this->expiration = $expiration;

		$this->production = $production;

		$this->refreshToken = $refreshToken;
	}

	/**
	 * Perform an AccessTokenRequest
	 * 
	 * @param string $clientId
	 * @param string $privateKey
	 * @param string $issuer
	 * @param \RevolutPHP\Auth\AuthToken $authToken
	 * 
	 * @return \RevolutPHP\Auth\AccessToken
	 */
	public static function request(string $clientId, string $privateKey, string $issuer, AuthToken $authToken, bool $production = true)
	{
		$request = new AccessTokenRequest($clientId, $privateKey, $issuer, $authToken, $production);

		return $request->get();
	}
}
```
### AccessTokenRequest
Uses GuzzleHttp to exchange an \Auth\AuthToken for an access token.
The AccessTokenRequest accepts either a RefreshToken or AuthrizationToken as its fourth parameter.

You can new up this class directly, or use the static 'request' method on AccessToken, it is the same.

```php

namespace RevolutPHP\Auth;

use RevolutPHP\Client as RevolutClient;
use GuzzleHttp\Client as GuzzleClient;
use GuzzleHttp\Exception\BadResponseException;

class AccessTokenRequest
{
	/**
	 * The auth endpoint
	 */
	const ENDPOINT = '/auth/token';

	private $clientId;

	private $privateKey;

	private $authToken;

	private $issuer;

	private $production;

	/**
	 * @param string $clientId The client ID
	 * @param string $privateKey The path to the private key file
	 * @param string $issuer
	 * @param \RevolutPHP\Auth\AuthToken $authToken
	 * @param bool $production
	 */
	public function __construct(string $clientId, string $privateKey, string $issuer, AuthToken $authToken, bool $production = false)
	{
		$this->clientId = $clientId;

		$this->privateKey = $privateKey;

		$this->issuer = $issuer;

		$this->authToken = $authToken;

		$this->production = $production;
	}

	/**
	 * Get a new access token instance
	 * 
	 * @return \RevolutPHP\Auth\AccessToken
	 */
	public function get()
	{
		$httpClient = new GuzzleClient();

		try {

			$response = $httpClient->request('POST', $this->url(), ['form_params' => $this->formParams()]);

		} catch(BadResponseException $error) {

			// handle your auth errors 
		}
		
		$data = json_decode($response->getBody(), true);

		$exp = $this->expirationTime($data['expires_in']);

		$refreshToken = $data['refresh_token'] ?? null;

		return new AccessToken($data['access_token'], $exp, $this->production, $refreshToken);
	}

	/**
	* @return int The timestamp
	 */
	private function expirationTime($seconds)
	{
		return time() + $seconds;
	}

	/**
	 * @return string
	 */
	private function url()
	{
		$url = $this->production
			? RevolutClient::REVOLUT_PRODUCTION_ENDPOINT
			: RevolutClient::REVOLUT_SANDBOX_ENDPOINT;
		
		return $url . RevolutClient::REVOLUT_API_VERSION . self::ENDPOINT;
	}

	/**
	 * @return array
	 */
	private function formParams()
	{
		$params = [
			'client_id' => $this->clientId,
			'client_assertion_type' => ClientAssertion::TYPE,
			'client_assertion' => $this->clientAssertion()->make(),
		];

		return array_merge($params, $this->authParams()); // build and add the missing authorisation code and grant type to the params 
	}

	/**
	 * Build the authentication params
	 * 
	 * @return array
	 */
	private function authParams()
	{
		return [
			$this->authToken->name() => $this->authToken->token,
			'grant_type' => $this->authToken->grantType(),
		];
	}

	/**
	 * Get the client assertion
	 * 
	 * @return \RevolutPHP\Auth\ClientAssertion
	 */
	private function clientAssertion()
	{
		return new ClientAssertion($this->clientId, $this->privateKey, $this->issuer);
	}
}
```
### ClientAssertion

Use firebase/php-jwt to create a JWT.

```php

namespace RevolutPHP\Auth;

use Firebase\JWT\JWT;

class ClientAssertion
{
	const TYPE = 'urn:ietf:params:oauth:client-assertion-type:jwt-bearer';

	const AUDIENCE = 'https://revolut.com';

	const ALGORYTHM = 'RS256';

	private $subject;

	private $privateKey;

	private $issuer;

	protected $expiration;

	public function __construct(string $clientId, string $privateKey, string $issuer, $expiration = null)
	{
		$this->subject = $clientId;

		$this->privateKey = $privateKey;

		$this->issuer = $issuer;

		$this->expiration = $expiration ?? time() + 60 * 5;
	}

	/**
	 * Create the Client Assertion
	 * 
	 * @return string The assertion string
	 */
	public function make()
	{
		return JWT::encode($this->payload(), $this->privateKey(), self::ALGORYTHM);
	}

	/**
	 * Get the contents of the private key
	 * 
	 * @return string|false
	 */
	private function privateKey()
	{
		return file_get_contents($this->privateKey);
	}

	/**
	 * Build the payload for the JWT
	 * 
	 * @return array The payload
	 */
	private function payload()
	{
		return [
			'iss' => $this->issuer, 
			'sub' => $this->subject,
			'exp' => $this->expiration,
			'aud' => self::AUDIENCE,
		];
	}
}
```

### How to use it

When you first authorize your application to use Revolut's api, you will receive an authorization code.
Simply new up a RevolutPHP\Auth\AuthorizationToken with the authorization code, and exchange it for an access token:
```php
$token = new AuthorizationToken($code);

$accessToken = AccessToken::request('clientId', '/path/to/privateKey', 'mysite.com', $token);
```
At this point, you can use the access token to new up a \RevolutPHP\Client.
When retrieving an access token via an authorization code, the \RevolutPHP\Auth\AccessToken will also include a refresh token, 
you should store this somewhere safe, as you will be using this refresh token to generate new access tokens.

access the refresh token on the AccessToken:
```php
$refreshToken = $accessToken->refreshToken;
```

To use the refresh token to generate a new access token, again, just new up a RevolutPHP\Auth\RefreshToken with the refresh token, and request a new access token;
```php
$token = new RefreshToken($refreshToken);

$accessToken = AccessToken::request('clientId', '/path/to/privateKey', 'mysite.com', $token);
```

**Comment by sverraest** [comment]
Thanks for the great work @tbclla and @vdbelt.

The PR has been merged to master, I'll leave it in master for now and try to get some of the other improvements in before tagging a new release. 

In your dependencies you can change the version to dev-master to use the new flow and then later the new stable release version once tagged.

**Comment by sverraest** [comment]
Released in v1.2.0
